### PR TITLE
Support Vault KV v2 backend. Connected to #1331.

### PR DESCRIPTION
**Pull request by michael2m**
Signed-off-by: Meijer <michaelmeijer@outlook.com>

### What does this PR do?
- Support secrets of Vault's KV v2 backend which is incompatible with the current implementation in secretless-broker.
- New behavior is implemented with backwards compatability to current (old) behavior.

### What ticket does this PR close?
Connected to #1331 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [x] This PR does not require updating any documentation, or
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

### CODE REVIEW

**Comment by izgeri** [comment]
Thanks for submitting this @michael2m! I'll have the team review the code within the next couple of business days.

Given that this increases the complexity of the `vault` provider configuration, would you be able to help us in writing a little documentation for this provider? We could add a README.md file to the `/internal/providers/vault/` folder that has some info about how to configure Secretless with vault. That would include something like:
```markdown
# HashiCorp Vault Provider

## Configuration
<Required steps to configure Secretless to authenticate with HCV>

## Using the `vault` provider
<Instructions for referencing credentials from vault using this provider, including examples when using `kv` or `kv-v2`>
```

Even if we just added a godoc-style comment to the top of the `provider.go` file so that it would have some info here: https://godoc.org/github.com/cyberark/secretless-broker/internal/providers/vault, that would be good - especially since there's now more than one way to use this provider.

I would like to add a CHANGELOG entry for this, too - we provide [some guidelines](https://github.com/cyberark/community/blob/master/Conjur/CONTRIBUTING.md#changelog-guidelines) for writing good change log messages. If you'd prefer, we can add a commit later that adds a message for your changes. I might just write something like this under `Added`:
```
- The `vault` provider now supports loading secrets from the KV Secrets Engine - Version 2. To reference secrets using this version, preface the secret path with `kv-v2:` in the Secretless configuration. [cyberark/secretless-broker#1331](<link>)
```

**Comment by michael2m** [comment]
You're totally right @izgeri. I will add some godoc comments. Just wondering about the docs though. It seems there was a documentation page once for the Vault provider, however it links to a non-existing page in the Conjur docs (secretless-broker/docs/docs/reference/providers/vault.md). I would prefer proper documentation, rather than a README somewhere in code.

**Comment by izgeri** [comment]
@michael2m this provider is still considered [community level](https://github.com/cyberark/community/blob/master/Conjur/conventions/certification-levels.md), which we document in-repo. You can see an example [here](https://github.com/cyberark/secretless-broker/tree/master/internal/plugin/connectors/http/generic) for the generic HTTP connector.

I do have it as a TODO to make sure all community providers / connectors have this minimal documentation, and that they're linked from the project README so they're easier to find. Whether we choose to augment the godocs or add a README here, I'm fine either way - either one is linkable from the README.

Hope that makes sense, but if you have any additional questions please lmk.

File: internal/providers/vault/README.md
Comment (sgnn7) [inline internal/providers/vault/README.md:11]: You should use a code block for the config here probably

File: internal/providers/vault/README.md
Comment (sgnn7) [inline internal/providers/vault/README.md:14]: Here we should mention which ones we support.

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:61]: We should raise an explicit error here if the backend ID is not supported

**Review by sgnn7 — CHANGES_REQUESTED** [review]
@michael2m Great work and thank you for your contribution! There's just a handful of things that may need a bit fixing. Also, your branch contains 2 signed commits (wanted) and an unwanted merge commit which will need a fix before we get this approved.

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go:134]: Will fix

**Comment by izgeri** [comment]
@michael2m I see that you're waiting on feedback on this PR - apologies for the delay, we've been pushing hard to get a new release out for another project. I'll make sure we get back to you on this by tomorrow. (cc @sgnn7)

File: internal/providers/vault/provider.go
Comment (sgnn7) [reply internal/providers/vault/provider.go:134]: > Aren't named return values a common practice ?

Yes, a very common one but the problem is that for maintenance, they make things more difficult than the alternative. The problem is that as someone is going down the flow of the method reading the code, they will inevitably either miss the fact that some variable is a side-effect return value and/or they will need to go back up to the function definition to see its type. This kind of implicit return effect as well as back-and-forth reading has a real penalty for maintenance which we would like to avoid.

File: internal/providers/vault/provider.go
Comment (sgnn7) [reply internal/providers/vault/provider.go:57]: > Using kv: prefix would break existing behavior. I advice against it.

Can you elaborate on this (I'm a bit newish on the Vault API specifics)? How is a generic secret fetch different from KVv1?

> Keep in mind that 99% of the code is equal. It's just the unwrapping of the KV v2 response that is different. I don't think extensibility is improved by factoring out into kv and kv-v2 variants.

I'd disagree here to an extent. Even if we ignore the extensibility part which I for sure should have elaborated on more, there's a bit of a problem. Whenever multiple logic paths are embedded this way in a codebase, the mental overhead increases for the maintainer to know how they intertwine in the flow as they are parsing the code and the test matrix grows by orders of magnitude for each branch in the code. I think some of this pattern is generally described by [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) and more often than not it has held accurate in my experience.

File: internal/providers/vault/README.md
Comment (sgnn7) [reply internal/providers/vault/README.md:11]: The Vault provider is configured in the `secretless.yml`. Using
    ```yaml
    from: vault
    get: /path/to/secret/in/vault
    ```
    for the credentials in the ...

File: internal/providers/vault/README.md
Comment (sgnn7) [reply internal/providers/vault/README.md:14]: Oh I'm not suggesting adding additional support for the others, just that we mention which ones are supporting (or have confidence that they work)

File: internal/providers/vault/provider.go
Comment (sgnn7) [reply internal/providers/vault/provider.go:61]: This ties into my other comment about `kv:` prefix but at the very least yes we should trow an error if a backend is provided and we don't support it.

**Comment by sgnn7** [comment]
@michael2m Sorry for the delay on this and thank you for your patience! As @izgeri mentioned we've been really busy on getting our redesigned [Puppet integration](https://forge.puppet.com/cyberark/conjur) project released and we have more to work on that but I've had some time today to reply to your comments. Let me know if you have any questions!

PS: I think the PR will need a rebase and the git history will need a touch of cleanup.

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go:134]: Clear, will fix.

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go:57]: > Can you elaborate on this (I'm a bit newish on the Vault API specifics)? How is a generic secret fetch different from KVv1?

Vault looks like a file system where you read from / write to, with contents made up of JSON. The "old" implementation the Vault provider for Conjur expects the secret value as JSON field in the returned object. Which works for various secret engines supported by Vault, hence the generic secret fetch. A notable exception is the KVv2 secret engine, due to its additional wrapping of the secret value in another object within the response object. That's why it needs to be handled differently.

> ... single-responsibility principle ...
I will look into a better code design, will get back to you on this.

Comment (michael2m) [reply internal/providers/vault/provider.go]: I opted for an alternative design, in which the distinction between various Vault secret engines/backend doesn't appear as much as it did in my previous solution. I redesigned to have more expressive field navigation. It uses `path/to/secert#navigate.to.field` instead of prefixing with `kv-v2:`. Hence the field name has become field navigation, as if traversing through an object structure. This is both compatible to the current behavior as well as extends naturally to the KV v2 case.

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go]: Given new design, this should no longer be relevant.

**Comment by michael2m** [comment]
Redesigned the support for KV v2 in an alternative way (see review thread above). Also cleaned up commit history rebasing the pull request off master.

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:102]: You might be able to make this a bit cleaner with use of `:=`
```suggestion
	secret, err := p.Client.Logical().Read(path)
	if err != nil {
```

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:65]: I'm a bit confused why this value is declared outside of the `for` block if the check of it's value is only done immediately after this line here.

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:67]: This line needs word wrapping

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:73]:
```suggestion
	field := fields[len(fields)-1]
	value := obj[field]
```

I would also suggest that you name these vars a bit better. What field is this? What value is here?

File: internal/providers/vault/provider.go
Comment (sgnn7) [reply internal/providers/vault/provider.go:65]: I think this loop also needs a bit clearer logic about what it does either with more obvious code (preferred) or some really good comments

File: internal/providers/vault/provider.go
Comment (sgnn7) [inline internal/providers/vault/provider.go:75]: This (and other) comments should be capitalized

**Review by sgnn7 — CHANGES_REQUESTED** [review]
@michael2m Hey Michael, great work on this! I like the cleaner look of this quite a bit! I left some comments on it for you but it's mostly tiny nits.

File: CHANGELOG.md
Comment (sgnn7) [inline CHANGELOG.md:13]: Oops - forgot that this may need updating too since the implementation changed

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go:73]: Relates to your other remark on the same section of code.

File: internal/providers/vault/provider.go
Comment (michael2m) [reply internal/providers/vault/provider.go]: I rewrote / refactored. Effectively it takes an object and a path traversing through the object structure. Suppose obj = { "foo": { "bar": "qux" } } and fields = "foo.bar", then returned value is "qux". It navigates from outer to inner object and last field's value.

**Comment by michael2m** [comment]
@sgnn7 I hope the rewrite/refactor with comments helps you understand the solution.

File: internal/providers/vault/provider.go
Comment (sgnn7) [reply internal/providers/vault/provider.go]: Oh for sure - I wasn't confused about what you were trying to do. My comment was more to ensure others can easily understand what is done in the flow if they are reading this for the first time.

**Comment by sgnn7** [comment]
/approve 8d56e6354aa91bead3f4c3fc18a30be4d9ddb220

@izgeri Can you take a look too?

**Comment by sgnn7** [comment]
@michael2m This looks good to me - I just need one more approval to move the branch to the repo itself so that we can verify that tests pass. I'll keep this thread updated though.

**Comment by sgnn7** [comment]
Closing this branch as it is superseded by https://github.com/cyberark/secretless-broker/pull/1345

**Pull request closed without merging**
